packages/block-library/src/gallery/block.json

```diff
@@ -68,14 +31,39 @@
 			"type": "boolean",
 			"default": true
 		},
+		"linkTarget": {
+			"type": "string"
+		},
 		"linkTo": {
 			"type": "string"
 		},
 		"sizeSlug": {
 			"type": "string",
 			"default": "large"
+		},
+		"allowResize": {
+			"type": "boolean",
+			"default": false
+		},
+		"isGrouped": {
+			"type": "boolean",
+			"default": true
+		},
+		"imageCount": {
+			"type": "number",
+			"default": 0
+		},
+		"gutterSize": {
+			"type": "number"
 		}
 	},
+	"providesContext": {
+		"allowResize": "allowResize",
+		"isGrouped": "isGrouped",
+		"linkTo": "linkTo",
+		"linkTarget": "linkTarget",
+		"sizeSlug": "sizeSlug"
```

Comment:
I guess these are meant to be passed down to the images  to keep the same value for all images? Can't we start simpler instead and just let the images handle these. On the UI we can still target the children blocks and call their own `setAttributes` instead of setting the attribute on the Gallery block instead.

I'd love if this PR to be as impact less as possible on the image block.

Reply:
We have been through quite a few iterations of this, and this seems to be the least impactful way on the image block while still providing the UX that uses expect based on our first round of testing, ie. users want to:

- Add a gallery and expect that by default any changes to the gallery image settings will be applied to all the children
- If they edit a setting for an individual image, eg. give it a custom link, and then at the gallery level set the default link to attachment, they don't expect that to wipe the custom link in the single image, only affect images they haven't specifically changed.

While point one is possible by setting the child attributes from the parent, point two is difficult as it is impossible to know from the parent which of the children might have had their settings manually toggled. We spent a bit of time working with @jasmussen on the UX around this ... but open to suggestions on alternatives.

Reply:
> If they edit a setting for an individual image, eg. give it a custom link, and then at the gallery level set the default link to attachment, they don't expect that to wipe the custom link in the single image, only affect images they haven't specifically changed.

This is actually something I want to challenge. If you multiselect something (regardless of the context), say "text", and you remove "bold", it's going to make everything "bold" not just what was edited globally before. So it seems to me that overriding the single image choice should be the best way to go here.

Reply:
@jasmussen did you have any thoughts of this from our previous discussions about the new gallery settings. You may remember that we did have a prompt on the gallery settings to confirm that that user wanted to overwrite all child settings, but you saw that as an antipattern, so we moved to moved tp the child images only applying gallery settings if they hadn't been individually set.

I see it being a gotcha for users who may set custom urls on a large number of images, and then maybe toggle the linkTo select at the gallery level maybe thinking it will only affect new images but it in fact wipes all their custom settings. 

Very happy to discuss the options and approaches further though.

Reply:
I definitely think any blanket overwriting of custom urls on inner image blocks would be a bad experience for our users. Not only would it be unexpected but they might not even notice it right away, increasing the frustration. 

Perhaps the distinction here is that the user isn't actively multi-selecting images. Therefore they aren't making the choice to alter those inner images directly.

Reply:
It's a key discussion, with trade-offs on both sides. Thanks for the ping again, and thanks also for refreshing my memory:

> You may remember that we did have a prompt on the gallery settings to confirm that that user wanted to overwrite all child settings, but you saw that as an antipattern, so we moved to moved tp the child images only applying gallery settings if they hadn't been individually set

Yes, moving the logic to the image itself seemed like the best alternative to a "confirm" prompt.

On the flipside, Riad makes a good point:

> If you multiselect something (regardless of the context), say "text", and you remove "bold", it's going to make everything "bold" not just what was edited globally before.

This is a "destructive" change in the same way as overriding link destinations on images in a gallery would be. The key difference is the bold toggle is very visual, whereas the gallery change is mostly invisible. You wouldn't necessarily know that this one image in the corner, which you linked to a rickroll video, had its URL changed.

But to Riad's point again — maybe you should have assumed this because it's the same as the "bold" behavior. And maybe it _is_ correct behavior after all. And maybe linking just one image is an edgecase not worth adding undue complexity for. And maybe we should embrace that there are highly visible undo/redo buttons in the toolbar.

Because I argued for the opposite, I refreshed the branch and took it for a spin to get a feel for it again:

![gallery](https://user-images.githubusercontent.com/1204802/107747394-a1bb1580-6d17-11eb-969c-5c814cfba286.gif)

So good. The fact that I can click an image to select it, press Delete to remove it directly. That I can rearrange these? This is all so good. And the "select parent" change really benefits this.

I tend to still think it would be nice to inform the user, somehow, that these gallery-level link changes might override that carefully constructed URL. But given we have undo, and version history, I do see where Riad is coming from.

Could a compromise be to go with Riad's approach, but to throw up a snackbar notice when you make a URL change? "All images in gallery now link to Media File"?

Reply:
@jasmussen, @youknowriad I have refactored in line with your suggestion Joen and the update now is handled by the gallery with a snackbar to indicate that all images were affected. The only issue is that each image update is a separate entry in the undo stack, so to undo the change in gallery with 3 images you need to hit undo 3 times. Do either of you know if there is a way to make batch a series of changes so they can be undone in one go? I ran out of time to explore the code for any examples of this today.

Reply:
Actually it looks like updating the block attributes by passing an array of block ids to updateBlockAttributes instead of looping through the innerblock array and setting for each block batches up the undos, so will look at refactoring to that tomorrow ... the only thing is it doesn't undo the setting in the gallery side panel, so might still be useful to know if there is a way to specify the beginning and end of a transaction in the undo queue

Reply:
> so to undo the change in gallery with 3 images you need to hit undo 3 times. Do either of you know if there is a way to batch a series of changes so they can be undone in one go?

I think we do support multiselection of blocks and changing attributes for all these blocks at the same time, so it seems both these two things should work the same way. I'm not sure whether this use-case creates a single undo or not though. If not, it should be as easy as having a new action on the block-editor package to update all the blocks in one go.

Reply:
> I think we do support multiselection of blocks and changing attributes for all these blocks at the same time

Yeh, it turns out you can update multiple blocks with `updateBlockAttributes` by passing an array of clientIds, and it gets added as a single event in undo stack ... but all the attributes have to be the same, and unfortunately with the likes of linkTo changes the href attrib is different for every nested image block so `updateBlockAttributes` has to be run separately for each block.

I will take a look at the option of adding a new action to the block-editor package tomorrow ... I assume that would be better as a standalone PR? New are for me so not sure how far I will get with that.

Reply:
@youknowriad it looks like this could be as easy as extending `updateBlockAttributes` to allow a clientId keyed object of attributes as an alternative to a flat object of attributes. I have a prototype of this at https://github.com/WordPress/gutenberg/pull/29099, and this works when tested with gallery pr to allow all gallery level attribute changes to be undone in one go. If you think this looks like a way forward I will do some more testing and add some additional unit tests to the PR.